src/Utils/dialogUtils.ts

```diff
@@ -133,3 +133,80 @@ export function getBestAction(scoreResponse: CLM.ScoreResponse, allActions: CLM.
     }
     return best
 }
+
+function doesScorerStepMatch(scorerStep1: CLM.TrainScorerStep, scorerStep2: CLM.TrainScorerStep): boolean {
+    if (scorerStep1.labelAction !== scorerStep2.labelAction) {
+        return false
+    }
+    if (scorerStep1.input.filledEntities.length !== scorerStep2.input.filledEntities.length) {
+        return false
+    }
+    let entityIndex = 0
+    while (entityIndex < scorerStep1.input.filledEntities.length) {
+        let entity1Id = scorerStep1.input.filledEntities[entityIndex].entityId
+        if (!scorerStep2.input.filledEntities.find(fe => fe.entityId === entity1Id)) {
+            return false
+        }
+        entityIndex = entityIndex + 1
+    }
+    return true
+}
+
+function doesRoundMatch(round1: CLM.TrainRound, round2: CLM.TrainRound, isLastRound: boolean): boolean {
+    
+    // If one has scorer steps and the other doesn't, only ok, on last round
+    if (round1.scorerSteps && !round2.scorerSteps ||
+        !round1.scorerSteps && round2.scorerSteps) {
+            return isLastRound
+        }
+    // If they both don't have scorer steps
+    if (!round1.scorerSteps && !round2.scorerSteps) {
+        return true
+    }
+
+    // Test that scorer steps match
+    const maxSteps = Math.max(round1.scorerSteps.length, round2.scorerSteps.length)
+    let stepIndex = 0
+    while (stepIndex < maxSteps) {
+        const scorerStep1 = round1.scorerSteps[stepIndex]
+        const scorerStep2 = round2.scorerSteps[stepIndex]
+        // If the last round it's ok that round has extra scorer steps
+        const extraScorerStep = (scorerStep1 && !scorerStep2) || (!scorerStep1 && scorerStep2)
+        if (extraScorerStep) {
+            return isLastRound
+        }
+        if (!doesScorerStepMatch(scorerStep1, scorerStep2)) {
+            return false
+        }
+        stepIndex = stepIndex + 1
+    }
+    return true
+}
+
+export function doesTrainDialogMatch(trainDialog1: CLM.TrainDialog, trainDialog2: CLM.TrainDialog): boolean {
+    const maxRounds = Math.max(trainDialog1.rounds.length, trainDialog2.rounds.length)
+    let roundIndex = 0
+    while (roundIndex < maxRounds) { 
```

Comment:
Seems like this is equivalent to an `zip` and `every` call which, in my opinion, more clearly explains what it's doing.
`zip` is common function that pairs elements of array limited by smaller array together (pairing rounds)
`every` is common function that say every round must match

Something like:
```
return zip(trainDialog1.rounds, trainDialog2.rounds)
  .every(({ round1, round2 }, roundIndex, roundPairs) => {
     const isLastRound = (roundIndex === roundPairs.length - 1)
     return doesRoundMatch(round1, round2, isLastRound)
  })
```

`zip` isn't standard javascript yet though so many libraries provide it